### DISCUSSION TITLE
How to disable default paragraph tool

### DISCUSSION THREAD

**Discussion by ThemeMetro** [codex-team/editor.js#1897] [Q&A]
The question.

How to disable default paragraph tool?

**Comment by woren951** [comment]
You can specify default Tool.
 
From [types/configs/editor-config.d.ts](https://github.com/codex-team/editor.js/blob/65b86c92038694a80e6311377b510be439806863/types/configs/editor-config.d.ts#L28)

```
  /**
   * This Tool will be used as default
   * Name should be equal to one of Tool`s keys of passed tools
   * If not specified, Paragraph Tool will be used
   */
  defaultBlock?: string;
```

**Comment by ThemeMetro** [comment]
> You can specify default Tool.
> 
> From [types/configs/editor-config.d.ts](https://github.com/codex-team/editor.js/blob/65b86c92038694a80e6311377b510be439806863/types/configs/editor-config.d.ts#L28)
> 
> ```
>   /**
>    * This Tool will be used as default
>    * Name should be equal to one of Tool`s keys of passed tools
>    * If not specified, Paragraph Tool will be used
>    */
>   defaultBlock?: string;
> ```

but after this the toolbar option is not available to select another tool/block, that's why want disable this permanently.

**Comment by loic-bellinger** [comment]
I use `defaultBlock: 'myParagraph',` but the 'paragraph' tool coming with EditorJS is still being used. I don't understand why. Could you please help? It's like it cannot be overriden.

**Comment by haupv** [comment]
Hi all,
I've got the same issue. Is there any solution, please?!

**Comment by nberongoy** [comment]
I added defaultBlock but still text is showing in the toolbar is there a way to hide it ?

**Reply by nberongoy** [reply]
One solution I found was via css 

```
 [data-item-name="paragraph"] {
        display: none;
      }
```

**Comment by misterjulo** [comment]
For those who come across this problem

just set paragraph to false like this

``` javascript
import MyCustomBlock from './plugins/MyCustomBlock'
new EditorJS({
      holder: 'editorjs',
      defaultBlock: 'customBlock'
      tools: {
        paragraph: false,
        customBlock: MyCustomBlock
      }
})
```

**Reply by ARH3** [reply]
@misterjulo when I do this, I get the following (yellow) error in the console:

`Paste handling for «paragraph» Tool hasn't been set up because of the error TypeError: this.constructable is not a constructor
`

This also messes up the Undo plugin, I've noticed.

Any idea how to disable Paragraph without throwing the error?

**Reply by ThemeMetro** [reply]
Is there nay update in this issue?

**Reply by ARH3** [reply]
Not that I've seen :(

**Reply by ThemeMetro** [reply]
> Not that I've seen :(

It working fine in 2.15.0 but when i am trying to update by latest version getting issue,

**Reply by tiagozip** [reply]
Thank you so much, this works perfectly!

**Reply by digisomni** [reply]
This doesn't quite work, it throws an error about "readonly" mode not being available on "paragraph: false".

**Comment by tiagozip** [comment]
Maybe it's a problem with your paragraph implementation. For me, I gave up
because there is no toolbar support for custom paragraphs.

A sábado, 30/12/2023, 4:28 da tarde, Kalila ***@***.***>
escreveu:

**Comment by MuhammadAhmedev** [comment]
Facing issue when i false paragraph then I can't able to get block content 

complete component :
import React, { useRef, useEffect, useState } from "react";
import TitleTool from "./tools/BigTitle/TitleTool";
import ButtonTool from "./tools/Button/ButtonTool";

export default function Editor({
  initialContent,
  data,
  onChange,
}: {
  initialContent?: any;
  data?: any;
  onChange?:any
}) {
  const [editorLoaded, setEditorLoaded] = useState(false);
  const editorRef = useRef(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const initializeEditor = async () => {
      const EditorJS = (await import("@editorjs/editorjs")).default;
      const Header = (await import("@editorjs/header")).default;
      const Table = (await import("@editorjs/table")).default;
      const List = (await import("@editorjs/list")).default;
      const LinkTool = (await import("@editorjs/link")).default;
      const RawTool = (await import("@editorjs/raw")).default;
      const SimpleImage = (await import("@editorjs/simple-image")).default;
      const Checklist = (await import("@editorjs/checklist")).default;
      const Embed = (await import("@editorjs/embed")).default;
      const Quote = (await import("@editorjs/quote")).default;
      const Paragraph = (await import("@editorjs/paragraph")).default;
    
      const tools = {
        header: Header,
        table: Table,
        list: List,
        linkTool: LinkTool,
        raw: RawTool,
        image: SimpleImage,
        checklist: Checklist,
        embed: Embed,
        paragraph: false,
        // paragraph: Paragraph,
        quote: {
          class: Quote,
          config: {
            quotePlaceholder: "Quote here...",
            captionPlaceholder: "Author...",
          },
        },
        // title: TitleTool,
        // buttonTool: ButtonTool,
      };

      const editorInstance = new EditorJS({
        holder: "editorjs",
        autofocus: true,
        // paragraph: false,
        placeholder: "Let`s write an awesome story!",
        inlineToolbar: true,
        tools,
        // defaultBlock: "Text",
        data: initialContent
          ? {
              time: initialContent.time,
              blocks: initialContent.block.blocks,
            }
          : ({} as any),
      } as any);

      editorRef.current = editorInstance;
      setEditorLoaded(true);
    };

    initializeEditor();

    return () => {
      if (editorRef.current) {
        editorRef.current.destroy();
      }
    };
  }, [initialContent]);

  const save = async () => {
    try {
      setLoading(true);
      if (editorRef.current) {
        const outputData = await editorRef.current.save();
        console.log('outputData', outputData)
        // Save data to the database
        return
        try {
          const saveInDb = await fetch("/api/post", {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(outputData),
          });
          console.log("Save In DB Response:", saveInDb);
        } catch (error) {
          console.error(
            "Error while saving editor content to the database:",
            error
          );
        }
      } else {
        console.log("Editor is not initialized");
      }
    } catch (error) {
      console.error("Error while saving editor content:", error);
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="flex justify-center flex-col mx-auto w-[70%]">
      <div
        id="editorjs"
        className="min-h-[50vh] min-w-[80%] border-[1px] rounded-md"
        style={{ border: "1px solid lightgray", padding: "20px" }}
      />
      <div className="flex justify-center my-[20px]">
        <button
          onClick={save}
          disabled={loading}
          className="bg-red-700 px-8 cursor-pointer py-3 w-fit rounded-[10px] text-white font-bold text-[14px]"
        >
          {loading ? "Loading..." : "Submit"}
        </button>
      </div>
    </div>
  );
}



output :
![image](https://github.com/codex-team/editor.js/assets/163818554/180e388d-82db-4a7e-b8c7-aa2959c8173b)

**Comment by ThemeMetro** [comment]
I think, instead of relying on the paragraph block in Editor.js, it should use an empty block as the default.

**Comment by hirnlego** [comment]
Solution for me was to define a custom tool:

```
class MyParagraph {
    render()
    {
        return null;
    }

    save()
    {
        return null;
    }
}
```

and then replacing the default in the configuration:

```
tools: {
    paragraph: MyParagraph,
    // Other tools
}
```

**Comment by davegurnell** [comment]
It looks like the paragraph tool appears in a list of core tools that are always added to the user-supplied list if the user doesn't explicitly override them as suggested above. The code that does this [is here](https://github.com/codex-team/editor.js/blob/next/src/components/modules/tools.ts#L196-L200) in `src/components/modules/tools.js`.

Here's a [yarn-patch](https://yarnpkg.com/cli/patch) compatible patch file that does the job on the ESM build that's shipped in version 2.30.8 on NPM. It's not an ideal solution but it's effective as a workaround for me at the moment:

```diff
diff --git a/dist/editorjs.mjs b/dist/editorjs.mjs
index 5a44e08549f1547de8550e38f2914dfab8d657ec..e773c1a800e0a32d38ef22cd4b716fa99e47698d 100644
--- a/dist/editorjs.mjs
+++ b/dist/editorjs.mjs
@@ -10391,11 +10391,11 @@ class Wn extends E {
         class: mo,
         isInternal: !0
       },
-      paragraph: {
-        class: fo,
-        inlineToolbar: !0,
-        isInternal: !0
-      },
+      // paragraph: {
+      //   class: fo,
+      //   inlineToolbar: !0,
+      //   isInternal: !0
+      // },
       stub: {
         class: jn,
         isInternal: !0
```

There are a couple of ways around this if the authors are willing to accept them as a PR:

- One possible approach would be to explicitly allow disabling of built-in tools by passing `null` or `false` in the config parameter as misterjulo suggested earlier in this thread (with an appropriate disclaimer that this might break things).

- Another possibility would be to add a second flag to disable the built-in tools (again, with an appropriate disclaimer).

**Reply by villeneuve-michael** [reply]
Any news on this? I'm currently using a CSS solution but it's not ideal! I would really love to either pass paragraph: false or simply not show it if we're overriding the default component (it could still be imported manually).

I don't mind working on the PR, would @neSpecc and the team be interested?
